src/Cryptography/Sodium/Bindings/Hashing.hs

```diff
@@ -0,0 +1,178 @@
+{-# LANGUAGE CApiFFI #-}
+{-# LANGUAGE Trustworthy #-}
+
+-- |
+--
+-- Module: Cryptography.Sodium.Bindings.Hashing
+-- Description: Direct bindings to the generic hashing primitives of Libsodium.
+-- Copyright: (C) Hécate Moonlight 2022
+-- License: BSD-3-Clause
+-- Maintainer: The Haskell Cryptography Group
+-- Stability: Stable
+-- Portability: GHC only
+module Cryptography.Sodium.Bindings.Hashing
+  ( -- * Introduction
+    -- $introduction
+
+    -- * Operations
+    CryptoGenericHashState,
+    allocateGenericHashState,
+    allocateGenericHashStateOfSize,
+    cryptoGenericHashInit,
+    cryptoGenericHashUpdate,
+    cryptoGenericHashFinal,
+
+    -- * Constants
+    cryptoGenericHashBytes,
+    cryptoGenericHashBytesMin,
+    cryptoGenericHashBytesMax,
+    cryptoGenericHashKeyBytes,
+    cryptoGenericHashKeyBytesMin,
+    cryptoGenericHashKeyBytesMax,
+  )
+where
+
+import Foreign (Ptr, mallocBytes)
+import Foreign.C (CInt (..), CSize (..), CUChar (..), CULLong (..))
+
+-- $introduction
+-- This API computes a fixed-length fingerprint for an arbitrarily long message.
+-- It is backed by the BLAKE2b algorithm.
+--
+-- Sample use cases:
+--
+--   * File integrity checking
+--   * Creating unique identifiers to index arbitrarily long data
+--
+-- ⚠ Do not use this API module to hash passwords!
+--
+-- Whenever there is a @'Ptr' 'CryptoGenericHashState'@, it must point to enough memory
+-- to hold the hash state.
+-- This is at least 'cryptoGenericHashBytesMin', at most
+-- 'cryptoGenericHashBytesMax', and should typically be 'cryptoGenericHashBytes'.
+-- It is the caller's responsibility to ensure that this holds.
+
+-- | Opaque tag representing the hash state struct @crypto_generichash_state@ used by the C API.
+--
+-- To obtain a 'CryptoGenericHashState', use 'allocateGenericHashState'.
+--
+-- @since 0.0.1.0
+data CryptoGenericHashState
+
+-- | This function allocates a 'CryptoGenericHashState' of size 'cryptoGenericHashBytes'.
+-- If you want more control over the size of the hash state, use 'allocateGenericHashStateOfSize'.
+--
+-- @since 0.0.1.0
+allocateGenericHashState :: IO (Ptr CryptoGenericHashState)
+allocateGenericHashState = allocateGenericHashStateOfSize cryptoGenericHashBytes
+
+-- | This function allocates a 'CryptoGenericHashState' of the desired size.
+--
+-- Use the following constants as parameter to this function:
+--
+--   * 'cryptoGenericHashBytesMin' (16U)
+--   * 'cryptoGenericHashBytes' (32U)
+--   * 'cryptoGenericHashBytesMax' (64U)
+--
+-- @since 0.0.1.0
+allocateGenericHashStateOfSize :: CSize -> IO (Ptr CryptoGenericHashState)
```

Comment:
I don't think this functionality should be provided here. You might want to use `allocaBytes` instead, or put it in a `ForeignPtr`.

Furthermore, the documentation is not entirely accurate: the min and max are _bounds_, so anything with size between them will also work.